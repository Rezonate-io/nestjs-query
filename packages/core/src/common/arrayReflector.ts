import { Reflector } from './reflector';
import { Class } from './class.type';
import { getClassMetadata, MetaValue } from './reflect.utils';

export class ArrayReflector extends Reflector {
  append<DTO, Data>(DTOClass: Class<DTO>, data: Data): void {
    const metadata = getClassMetadata<DTO, Data[]>(DTOClass, this.metaKey, false) ?? [];
    metadata.push(data);
    this.defineMetadata(metadata, DTOClass);
  }

  get<DTO, Data>(DTOClass: Class<DTO>, includeParents = false): MetaValue<Data[]> {
    return this.getMetadata(DTOClass, includeParents);
  }
}