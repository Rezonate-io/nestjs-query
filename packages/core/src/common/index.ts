export * from './class.type';
export * from './class.utils';
export * from './deep-partial.type';
export * from './misc.utils';
export * from './reflect.utils';
export { Reflector } from './reflector';
export { ValueReflector } from './valueReflector';
export { ArrayReflector } from './arrayReflector';
export { MapReflector } from './mapReflector';
