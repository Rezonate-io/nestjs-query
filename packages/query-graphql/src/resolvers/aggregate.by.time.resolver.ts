// eslint-disable-next-line max-classes-per-file
import { Args, ArgsType, Resolver } from '@nestjs/graphql';
import {
  AggregateByTimeResponse,
  AggregateQuery,
  AggregateResponse,
  Class,
  Filter,
  mergeFilter,
  QueryService,
} from '@rezonate/nestjs-query-core';
import omit from 'lodash.omit';

import { OperationGroup } from '../auth';
import { getDTONames } from '../common';
import {  AuthorizerFilter, ResolverQuery } from '../decorators';
import { AuthorizerInterceptor } from '../interceptors';
import { transformAndValidate } from './helpers';
import { BaseServiceResolver, ResolverClass, ServiceResolver } from './resolver.interface';
import { AggregateByTimeResponseType } from '../types/aggregate/aggregate-by-time-response.type';
import { AggregateByTimeArgsType } from '../types/aggregate/aggregate-by-time-args.type';
import { AggregateResolverOpts } from './aggregate.resolver';
import { AggregateByTimeQueryParam } from '../decorators/aggregate-by-time-query-param.decorator';

export interface AggregateByTimeResolver<DTO, QS extends QueryService<DTO, unknown, unknown>> extends ServiceResolver<DTO, QS> {
  aggregateByTime(
    filter: AggregateByTimeArgsType<DTO>,
    aggregateQuery: AggregateQuery<DTO>,
    authFilter?: Filter<DTO>
  ): Promise<AggregateResponse<DTO>[]>
}

/**
 * @internal
 * Mixin to add `aggregate` graphql endpoints.
 */
export const AggregateableByTime =
  <DTO, QS extends QueryService<DTO, unknown, unknown>>(DTOClass: Class<DTO>, opts?: AggregateResolverOpts) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): Class<AggregateByTimeResolver<DTO, QS>> & B => {
    if (!opts || !opts.enabled) {
      return BaseClass as never;
    }

    const { baseNameLower } = getDTONames(DTOClass);
    const commonResolverOpts = omit(opts, 'dtoName', 'one', 'many', 'QueryArgs', 'Connection');
    const queryName = `${baseNameLower}AggregateByTime`;
    const AR = AggregateByTimeResponseType(DTOClass);

    @ArgsType()
    class AA extends AggregateByTimeArgsType(DTOClass) {}

    @Resolver(() => AR, { isAbstract: true })
    class AggregateByTimeResolverBase extends BaseClass {
      @ResolverQuery(
        () => [AR],
        { name: queryName },
        commonResolverOpts,
        { interceptors: [AuthorizerInterceptor(DTOClass)] },
        opts ?? {},
      )
      async aggregateByTime(
        @Args({ type: () => AA }) args: AA,
        @AggregateByTimeQueryParam() query: AggregateQuery<DTO>,
        @AuthorizerFilter({
          operationGroup: OperationGroup.AGGREGATE,
          many: true,
        })
        authFilter?: Filter<DTO>,
      ): Promise<AggregateByTimeResponse<DTO>> {
        const qa = await transformAndValidate(AA, args);
        const filter = mergeFilter(qa.filter || {}, authFilter ?? {});

        return this.service.aggregateByTime(
          filter,
          query,
          args.field,
          args.from,
          args.to ?? new Date(),
          args.interval.count,
          args.interval.span,
          args.accumulate,
          qa.groupByLimit,
          opts.maxRowsForAggregate,
          opts.maxRowsForAggregateWithIndex,
          opts.limitAggregateByTableSize,
        );
      }
    }

    return AggregateByTimeResolverBase;
  };
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const AggregateByTimeResolver = <
  DTO,
  QS extends QueryService<DTO, unknown, unknown> = QueryService<DTO, unknown, unknown>,
>(
  DTOClass: Class<DTO>,
  opts?: AggregateResolverOpts,
): ResolverClass<DTO, QS, AggregateByTimeResolver<DTO, QS>> => AggregateableByTime(DTOClass, opts)(BaseServiceResolver);
