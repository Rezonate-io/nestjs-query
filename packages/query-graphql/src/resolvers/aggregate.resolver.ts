import { Args, ArgsType, Resolver } from '@nestjs/graphql'
import { AggregateQuery, AggregateResponse, Class, Filter, mergeFilter, QueryService } from '@rezonate/nestjs-query-core'
import omit from 'lodash.omit'

import { OperationGroup } from '../auth'
import { getDTONames } from '../common'
import { AggregateQueryParam, AuthorizerFilter, ResolverMethodOpts, ResolverQuery } from '../decorators'
import { AuthorizerInterceptor } from '../interceptors'
import { AggregateArgsType, AggregateResponseType } from '../types'
import { transformAndValidate } from './helpers'
import { BaseServiceResolver, ResolverClass, ServiceResolver } from './resolver.interface'

export type AggregateResolverOpts = {
  enabled?: boolean
  maxRowsForAggregate?: number
  maxRowsForAggregateWithIndex?: number
} & ResolverMethodOpts

export interface AggregateResolver<DTO, QS extends QueryService<DTO, unknown, unknown>> extends ServiceResolver<DTO, QS> {
  aggregate(
    filter: AggregateArgsType<DTO>,
    aggregateQuery: AggregateQuery<DTO>,
    authFilter?: Filter<DTO>
  ): Promise<AggregateResponse<DTO>[]>
}

/**
 * @internal
 * Mixin to add `aggregate` graphql endpoints.
 */
export const Aggregateable =
  <DTO, QS extends QueryService<DTO, unknown, unknown>>(DTOClass: Class<DTO>, opts?: AggregateResolverOpts) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): Class<AggregateResolver<DTO, QS>> & B => {
    if (!opts || !opts.enabled) {
      return BaseClass as never
    }

    const { baseNameLower, baseName } = getDTONames(DTOClass)
    const commonResolverOpts = omit(opts, 'dtoName', 'one', 'many', 'QueryArgs', 'Connection')
    const queryName = `${baseNameLower}Aggregate`
    const AR = AggregateResponseType(DTOClass)

    @ArgsType()
    class AA extends AggregateArgsType(DTOClass) {}

    @Resolver(() => AR, { isAbstract: true })
    class AggregateResolverBase extends BaseClass {
      @ResolverQuery(
        () => [AR],
        { name: queryName },
        commonResolverOpts,
        { interceptors: [AuthorizerInterceptor(DTOClass)] },
        opts ?? {}
      )
      async aggregate(
        @Args({ type: () => AA }) args: AA,
        @AggregateQueryParam() query: AggregateQuery<DTO>,
        @AuthorizerFilter({
          operationGroup: OperationGroup.AGGREGATE,
          many: true
        })
        authFilter?: Filter<DTO>
      ): Promise<AggregateResponse<DTO>[]> {
        const qa = await transformAndValidate(AA, args)
        const filter = mergeFilter(qa.filter || {}, authFilter ?? {})

        return this.service.aggregate(filter, query, qa.groupByLimit, opts.maxRowsForAggregate, opts.maxRowsForAggregateWithIndex)
      }
    }

    return AggregateResolverBase
  }
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const AggregateResolver = <DTO, QS extends QueryService<DTO, unknown, unknown> = QueryService<DTO, unknown, unknown>>(
  DTOClass: Class<DTO>,
  opts?: AggregateResolverOpts
): ResolverClass<DTO, QS, AggregateResolver<DTO, QS>> => Aggregateable(DTOClass, opts)(BaseServiceResolver)
