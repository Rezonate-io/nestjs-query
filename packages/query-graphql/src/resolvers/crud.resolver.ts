import { Resolver } from '@nestjs/graphql';
import { Class, DeepPartial, QueryService } from '@rezonate/nestjs-query-core';

import { mergeBaseResolverOpts } from '../common';
import { BaseResolverOptions } from '../decorators/resolver-method.decorator';
import { ConnectionOptions, PagingStrategies } from '../types';
import { AggregateableByTime } from './aggregate.by.time.resolver';
import { Aggregateable, AggregateResolver, AggregateResolverOpts } from './aggregate.resolver';
import { Creatable, CreateResolver, CreateResolverOpts } from './create.resolver';
import { Deletable, DeleteResolver, DeleteResolverOpts } from './delete.resolver';
import { Readable, ReadResolverFromOpts, ReadResolverOpts } from './read.resolver';
import { Referenceable, ReferenceResolverOpts } from './reference.resolver';
import { Relatable } from './relations';
import { RelatableOpts } from './relations/relations.resolver';
import { BaseServiceResolver, MergePagingStrategyOpts } from './resolver.interface';
import { Updateable, UpdateResolver, UpdateResolverOpts } from './update.resolver';

export interface CRUDResolverOpts<
	DTO,
	R extends ReadResolverOpts<DTO> = ReadResolverOpts<DTO>,
	PS extends PagingStrategies = PagingStrategies.CURSOR,
> extends BaseResolverOptions,
	Pick<ConnectionOptions, 'enableTotalCount'> {
	/**
	 * The DTO that should be used as input for create endpoints.
	 */
	CreateDTOClass?: Class<DeepPartial<DTO>>;
	/**
	 * The DTO that should be used as input for update endpoints.
	 */
	UpdateDTOClass?: Class<DeepPartial<DTO>>;
	enableSubscriptions?: boolean;
	pagingStrategy?: PS;
	enableAggregate?: boolean;
	create?: CreateResolverOpts<DTO>;
	read?: R;
	update?: UpdateResolverOpts<DTO>;
	delete?: DeleteResolverOpts<DTO>;
	referenceBy?: ReferenceResolverOpts;
	aggregate?: AggregateResolverOpts;
	limitAggregateByTableSize?: boolean;
}

export interface CRUDResolver<
	DTO,
	R extends ReadResolverOpts<DTO>,
	QS extends QueryService<DTO> = QueryService<DTO>,
> extends CreateResolver<DTO, QS>,
	ReadResolverFromOpts<DTO, R, QS>,
	UpdateResolver<DTO, QS>,
	DeleteResolver<DTO, QS>,
	AggregateResolver<DTO, QS> {
}

function extractRelatableOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): RelatableOpts {
	const { enableTotalCount, enableAggregate, aggregate } = opts;
	return mergeBaseResolverOpts<RelatableOpts>({ enableAggregate, enableTotalCount, ...aggregate }, opts);
}

function extractAggregateResolverOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): AggregateResolverOpts {
	const { enableAggregate, aggregate } = opts;
	return mergeBaseResolverOpts<AggregateResolverOpts>({ enabled: enableAggregate, ...aggregate }, opts);
}

function extractCreateResolverOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): CreateResolverOpts<DTO> {
	const { CreateDTOClass, enableSubscriptions, create } = opts;
	return mergeBaseResolverOpts<CreateResolverOpts<DTO>>({ CreateDTOClass, enableSubscriptions, ...create }, opts);
}

function extractReadResolverOpts<DTO, R extends ReadResolverOpts<DTO>, PS extends PagingStrategies>(
	opts: CRUDResolverOpts<DTO, R, PagingStrategies>,
): MergePagingStrategyOpts<DTO, R, PS> {
	const { enableTotalCount, pagingStrategy, read } = opts;
	return mergeBaseResolverOpts({
		enableTotalCount,
		pagingStrategy, ...read,
	} as MergePagingStrategyOpts<DTO, R, PS>, opts);
}

function extractUpdateResolverOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): UpdateResolverOpts<DTO> {
	const { UpdateDTOClass, enableSubscriptions, update } = opts;
	return mergeBaseResolverOpts<UpdateResolverOpts<DTO>>({ UpdateDTOClass, enableSubscriptions, ...update }, opts);
}

function extractDeleteResolverOpts<DTO>(
	opts: CRUDResolverOpts<DTO, ReadResolverOpts<DTO>, PagingStrategies>,
): DeleteResolverOpts<DTO> {
	const { enableSubscriptions, delete: deleteArgs } = opts;
	return mergeBaseResolverOpts<DeleteResolverOpts<DTO>>({ enableSubscriptions, ...deleteArgs }, opts);
}

/**
 * Factory to create a resolver that includes all CRUD methods from [[CreateResolver]], [[ReadResolver]],
 * [[UpdateResolver]], and [[DeleteResolver]].
 *
 * ```ts
 * import { CRUDResolver } from '@rezonate/nestjs-query-graphql';
 * import { Resolver } from '@nestjs/graphql';
 * import { TodoItemDTO } from './dto/todo-item.dto';
 * import { TodoItemService } from './todo-item.service';
 *
 * @Resolver()
 * export class TodoItemResolver extends CRUDResolver(TodoItemDTO) {
 *   constructor(readonly service: TodoItemService) {
 *     super(service);
 *   }
 * }
 * ```
 * @param DTOClass - The DTO Class that the resolver is for. All methods will use types derived from this class.
 * @param opts - Options to customize the resolver.
 */
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const CRUDResolver = <
	DTO,
	R extends ReadResolverOpts<DTO> = ReadResolverOpts<DTO>,
	PS extends PagingStrategies = PagingStrategies.CURSOR,
>(
	DTOClass: Class<DTO>,
	opts: CRUDResolverOpts<DTO, R, PS> = {},
):Class<CRUDResolver<DTO, R, any>> => {

	@Resolver(() => DTOClass, { isAbstract: true })
	class BaseResolver extends BaseServiceResolver<DTO, any> {
	}

	const Mixins = [] as ((cls: Class<unknown>) => Class<unknown>)[];

	Mixins.push(Referenceable(DTOClass, opts.referenceBy ?? {}));
	Mixins.push(Relatable(DTOClass, extractRelatableOpts(opts)));
	if (opts.aggregate?.disabled) Mixins.push(Aggregateable(DTOClass, extractAggregateResolverOpts(opts)));
	if (opts.aggregate?.disabled) Mixins.push(AggregateableByTime(DTOClass, extractAggregateResolverOpts(opts)));
	if (opts.create?.disabled) Mixins.push(Creatable(DTOClass, extractCreateResolverOpts(opts)));
	if (opts.read?.disabled) Mixins.push(Readable(DTOClass, extractReadResolverOpts(opts)));
	if (opts.update?.disabled) Mixins.push(Updateable(DTOClass, extractUpdateResolverOpts(opts)));
	if (opts.delete?.disabled) Mixins.push(Deletable(DTOClass, extractDeleteResolverOpts(opts)));

	return Mixins.reduce(
		(CurrResolver, action) => action(CurrResolver),
		BaseResolver,
	) as Class<CRUDResolver<DTO, R, any>>;
};
