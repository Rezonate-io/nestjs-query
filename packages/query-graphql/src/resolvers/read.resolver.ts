// eslint-disable-next-line max-classes-per-file
import { ArgsType, Resolver } from '@nestjs/graphql';
import { Class, Filter, mergeQuery, QueryService } from '@rezonate/nestjs-query-core';
import omit from 'lodash.omit';
import Papa from 'papaparse';

import { OperationGroup } from '../auth';
import { getDTONames } from '../common';
import { AuthorizerFilter, getQueryOptions, HookArgs, ResolverQuery } from '../decorators';
import { HookTypes } from '../hooks';
import { AuthorizerInterceptor, HookInterceptor } from '../interceptors';
import {
  ConnectionOptions,
  FilterType,
  FindOneArgsType,
  InferConnectionTypeFromStrategy,
  PagingStrategies,
  QueryArgsType,
  QueryArgsTypeOpts,
} from '../types';
import { CursorQueryArgsTypeOpts, QueryType, StaticQueryType } from '../types/query';
import { DEFAULT_QUERY_OPTS } from '../types/query/query-args';
import { BaseServiceResolver, ExtractPagingStrategy, ResolverClass, ResolverOpts, ServiceResolver } from './resolver.interface';

const QUERY_ARGS_TOKEN = Symbol('QUERY_ARGS_TOKEN');

export type ReadResolverFromOpts<
  DTO,
  Opts extends ReadResolverOpts<DTO>,
  QS extends QueryService<DTO, unknown, unknown>,
> = ReadResolver<DTO, ExtractPagingStrategy<DTO, Opts>, QS>;

export type ReadResolverOpts<DTO> = {
  QueryArgs?: StaticQueryType<DTO, PagingStrategies>
  idField?: keyof DTO
} & ResolverOpts &
  QueryArgsTypeOpts<DTO> &
  Pick<ConnectionOptions, 'enableTotalCount'>;

export interface ReadResolver<DTO, PS extends PagingStrategies, QS extends QueryService<DTO, unknown, unknown>>
  extends ServiceResolver<DTO, QS> {
  queryMany(
    query: QueryType<DTO, PagingStrategies>,
    authorizeFilter?: Filter<DTO>
  ): Promise<InferConnectionTypeFromStrategy<DTO, PS>>

  queryManyToCSV(query: QueryType<DTO, PagingStrategies>, authorizeFilter?: Filter<DTO>): Promise<string>

  queryManyIds(query: QueryType<DTO, PagingStrategies>, authorizeFilter?: Filter<DTO>): Promise<string[]>

  findById(id: FindOneArgsType, authorizeFilter?: Filter<DTO>): Promise<DTO | undefined>
}

export const getQueryArgs = <DTO>(DTOClass: Class<DTO>) => FilterType(DTOClass);

const serializeNestedObjects = (obj: Record<string, any>): Record<string, any> => {
  const result: Record<string, any> = {};
  Object.entries(obj).forEach(([key, value]) => {
    if (typeof value === 'object' && value !== null) {
      result[key] = JSON.stringify(value);
    } else {
      result[key] = value;
    }
  });
  return result;
};

/**
 * @internal
 * Mixin to add `read` graphql endpoints.
 */
export const Readable =
  <DTO, ReadOpts extends ReadResolverOpts<DTO>, QS extends QueryService<DTO, unknown, unknown>>(
    DTOClass: Class<DTO>,
    opts: ReadOpts,
  ) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): Class<ReadResolverFromOpts<DTO, ReadOpts, QS>> & B => {
    const { baseNameLower, pluralBaseNameLower, baseName } = getDTONames(DTOClass, opts);
    const readOneQueryName = opts.one?.name ?? baseNameLower;
    const readManyQueryName = opts.many?.name ?? pluralBaseNameLower;
    const { QueryArgs = QueryArgsType(DTOClass, { ...opts, connectionName: `${baseName}Connection` }) } = opts;
    const { ConnectionType } = QueryArgs;

    Reflect.defineMetadata(QUERY_ARGS_TOKEN, QueryArgs, DTOClass);

    const commonResolverOpts = omit(opts, 'dtoName', 'one', 'many', 'QueryArgs', 'Connection', 'withDeleted');

    @ArgsType()
    class QA extends QueryArgs {}

    @ArgsType()
    class FO extends FindOneArgsType(DTOClass) {}

    @Resolver(() => DTOClass, { isAbstract: true })
    class ReadResolverBase extends BaseClass {
      @ResolverQuery(
        () => DTOClass,
        { name: readOneQueryName, description: opts?.one?.description },
        commonResolverOpts,
        { interceptors: [HookInterceptor(HookTypes.BEFORE_FIND_ONE, DTOClass), AuthorizerInterceptor(DTOClass)] },
        opts.one ?? {},
      )
      async findById(
        @HookArgs() input: FO,
        @AuthorizerFilter({
          operationGroup: OperationGroup.READ,
          many: false,
        })
        authorizeFilter?: Filter<DTO>,
      ): Promise<DTO> {
        return this.service.getById(input.id, {
          filter: authorizeFilter,
          withDeleted: opts?.one?.withDeleted,
        });
      }

      @ResolverQuery(
        () => QueryArgs.ConnectionType.resolveType,
        { name: readManyQueryName, description: opts?.many?.description },
        commonResolverOpts,
        { interceptors: [HookInterceptor(HookTypes.BEFORE_QUERY_MANY, DTOClass), AuthorizerInterceptor(DTOClass)] },
        opts.many ?? {},
      )
      async queryMany(
        @HookArgs() query: QA,
        @AuthorizerFilter({
          operationGroup: OperationGroup.READ,
          many: true,
        })
        authorizeFilter?: Filter<DTO>,
      ): Promise<InstanceType<typeof ConnectionType>> {
        return ConnectionType.createFromPromise(
          (q) => this.service.query(q),
          mergeQuery(query, { filter: authorizeFilter }),
          (filter) => this.service.count(filter),
        );
      }

      @ResolverQuery(
        () => String,
        { name: `${readManyQueryName}CSV`, description: opts?.many?.description },
        commonResolverOpts,
        { interceptors: [HookInterceptor(HookTypes.BEFORE_QUERY_MANY, DTOClass), AuthorizerInterceptor(DTOClass)] },
        opts.many ?? {},
      )
      async queryManyToCSV(
        @HookArgs() query: QA,
        @AuthorizerFilter({
          operationGroup: OperationGroup.READ,
          many: true,
        })
        authorizeFilter?: Filter<DTO>,
      ) {
        const limitValue = getQueryOptions(DTOClass).CSVPageLimit || DEFAULT_QUERY_OPTS.CSVPageLimit;
        const res = await this.service.query({ ...mergeQuery(query, { filter: authorizeFilter }), paging: { ...query.paging, limit: limitValue } });
        return Papa.unparse(res.map((r) => serializeNestedObjects(r)));
      }

      @ResolverQuery(
        () => [String],
        { name: `${readOneQueryName}Ids`, description: opts?.many?.description },
        commonResolverOpts,
        {
          interceptors: [HookInterceptor(HookTypes.BEFORE_QUERY_MANY, DTOClass), AuthorizerInterceptor(DTOClass)],
          disabled: !opts.idField,
        },
        opts.many ?? {},
      )
      queryManyIds(
        @HookArgs() query: QA,
        @AuthorizerFilter({
          operationGroup: OperationGroup.READ,
          many: true,
        })
        authorizeFilter?: Filter<DTO>,
      ) {
        return this.service.queryIds(mergeQuery(query, { filter: authorizeFilter }), opts.idField);
      }
    }

    return ReadResolverBase as Class<ReadResolverFromOpts<DTO, ReadOpts, QS>> & B;
  };
// eslint-disable-next-line @typescript-eslint/no-redeclare -- intentional
export const ReadResolver = <
  DTO,
  ReadOpts extends ReadResolverOpts<DTO> = CursorQueryArgsTypeOpts<DTO>,
  QS extends QueryService<DTO, unknown, unknown> = QueryService<DTO, unknown, unknown>,
>(
  DTOClass: Class<DTO>,
  opts: ReadOpts = {} as ReadOpts,
): ResolverClass<DTO, QS, ReadResolverFromOpts<DTO, ReadOpts, QS>> => Readable(DTOClass, opts)(BaseServiceResolver);
