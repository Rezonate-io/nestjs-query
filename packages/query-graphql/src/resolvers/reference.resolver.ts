import { BadRequestException } from '@nestjs/common';
import { Resolver, ResolveReference } from '@nestjs/graphql';
import { Class, QueryService } from '@rezonate/nestjs-query-core';

import { getDTONames } from '../common';
import { RepresentationType } from '../federation';
import { BaseServiceResolver, ResolverClass, ServiceResolver } from './resolver.interface';

export interface ReferenceResolverOpts {
  key?: string
}

export interface ReferenceResolverType<DTO, QS extends QueryService<DTO> = QueryService<DTO>> extends ResolverClass<DTO, QS, ServiceResolver<DTO, QS>> {
	resolveReference(representation: RepresentationType):Promise<DTO>;
}

/**
 * @internal
 * Mixin to expose `resolveReference` for a DTO on the resolver.
 */
export const Referenceable =
  <DTO, QS extends QueryService<DTO>>(DTOClass: Class<DTO>, opts: ReferenceResolverOpts) =>
  <B extends Class<ServiceResolver<DTO, QS>>>(BaseClass: B): B & Class<ReferenceResolverType<DTO, QS>> => {
    if (!('key' in opts) || opts.key === undefined) {
      return BaseClass as B & Class<ReferenceResolverType<DTO, QS>>;
    }
    const { key } = opts;

    @Resolver(() => DTOClass, { isAbstract: true })
    class ResolveReferenceResolverBase extends BaseClass {
      @ResolveReference()
      async resolveReference(representation: RepresentationType): Promise<DTO> {
        const id = representation[key];
        if (id === undefined) {
          throw new BadRequestException(
            `Unable to resolve reference, missing required key ${key} for ${getDTONames(DTOClass).baseName}`,
          );
        }
        return this.service.getById(representation[key] as string | number);
      }
    }

    return ResolveReferenceResolverBase as B & Class<ReferenceResolverType<DTO, QS>>;
  };

export const ReferenceResolver = <DTO, QS extends QueryService<DTO> = QueryService<DTO>>(
  DTOClass: Class<DTO>,
  opts: ReferenceResolverOpts = {},
): Class<ReferenceResolverType<DTO, QS>> => Referenceable<DTO, QS>(DTOClass, opts)(BaseServiceResolver);
