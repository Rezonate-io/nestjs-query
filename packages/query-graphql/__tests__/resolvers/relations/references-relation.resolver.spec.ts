import { Query, Resolver } from '@nestjs/graphql';

import { ReferencesOpts, ReferencesRelationsResolver } from '../../../src/resolvers/relations';
import {
	createResolverFromNest,
	generateSchema,
	TestRelationDTO,
	TestResolverDTO,
	TestService,
} from '../../__fixtures__';
import { Inject } from '@nestjs/common';

@Resolver(() => TestResolverDTO)
class TestResolver extends ReferencesRelationsResolver(TestResolverDTO, {
  reference: { DTO: TestRelationDTO, keys: { id: 'stringField' } },
}) {
	constructor(@Inject() service: TestService) {
		super(service);
	}
}

interface FinalizedTestResolver {
	referenceReference(dto:TestResolverDTO):any
}

describe('ReferencesRelationMixin', () => {
  const expectResolverSDL = async (opts?: ReferencesOpts<TestResolverDTO>) => {
    @Resolver(() => TestResolverDTO)
    class TestSDLResolver extends ReferencesRelationsResolver(TestResolverDTO, opts ?? {}) {
      @Query(() => TestResolverDTO)
      test(): TestResolverDTO {
        return { id: '1', stringField: 'foo' };
      }
    }

    const schema = await generateSchema([TestSDLResolver]);
    expect(schema).toMatchSnapshot();
  };
  it('should not add reference methods if references empty', () => expectResolverSDL());

  it('should use the add the reference if provided', () =>
    expectResolverSDL({ reference: { DTO: TestRelationDTO, keys: { id: 'stringField' }, dtoName: 'Test' } }));

  it('should set the field to nullable if set to true new test', () =>
    expectResolverSDL({ reference: { DTO: TestRelationDTO, keys: { id: 'stringField' }, nullable: true } }));

  it('should return a references type from the passed in dto', async () => {
    const { resolver } = await createResolverFromNest<TestResolver, FinalizedTestResolver>(TestResolver);
    const dto: TestResolverDTO = {
      id: 'id-1',
      stringField: 'reference-id-1',
    };
    const result = await resolver.referenceReference(dto);
    return expect(result).toEqual({ __typename: 'Reference', id: dto.stringField });
  });
});
