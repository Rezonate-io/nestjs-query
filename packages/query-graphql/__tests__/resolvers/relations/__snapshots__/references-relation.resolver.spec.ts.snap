// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferencesRelationMixin should not add reference methods if references empty 1`] = `
"type TestResolverDTO {
  id: ID!
  stringField: String!
}

\\"\\"\\"Cursor for paging through collections\\"\\"\\"
scalar ConnectionCursor

\\"\\"\\"
Relative date, an Epoch time (number), ISO Date string or a relative date value in the form of \\"last-[N]-[minutes | hours | days | weeks | months | years]\\"
\\"\\"\\"
scalar RelativeDate

type Query {
  test: TestResolverDTO!
}"
`;

exports[`ReferencesRelationMixin should set the field to nullable if set to true new test 1`] = `
"type TestResolverDTO {
  id: ID!
  stringField: String!
  test: TestRelationDTO!
  reference: TestRelationDTO
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

\\"\\"\\"Cursor for paging through collections\\"\\"\\"
scalar ConnectionCursor

\\"\\"\\"
Relative date, an Epoch time (number), ISO Date string or a relative date value in the form of \\"last-[N]-[minutes | hours | days | weeks | months | years]\\"
\\"\\"\\"
scalar RelativeDate

type Query {
  test: TestResolverDTO!
}"
`;

exports[`ReferencesRelationMixin should use the add the reference if provided 1`] = `
"type TestResolverDTO {
  id: ID!
  stringField: String!
  test: TestRelationDTO!
}

type TestRelationDTO {
  id: ID!
  testResolverId: String!
}

\\"\\"\\"Cursor for paging through collections\\"\\"\\"
scalar ConnectionCursor

\\"\\"\\"
Relative date, an Epoch time (number), ISO Date string or a relative date value in the form of \\"last-[N]-[minutes | hours | days | weeks | months | years]\\"
\\"\\"\\"
scalar RelativeDate

type Query {
  test: TestResolverDTO!
}"
`;
